### PR TITLE
trunk variant for NLS problems

### PR DESCRIPTION
**Pull request by dpo**
@abelsiqueira There's quite a bit of repetition with `trunk` here. Do you see a more elegant way that would avoid redundancy?

### CODE REVIEW

**Comment by abelsiqueira** [comment]
I haven't looked at it in details, but it seems that the main difference is the step using the Hessian approximation of Gauss-Newton solved by least squares cg.
The objective function and gradients could be computed at the same time using `objgrad!`, so it would also avoid unnecessary computations. So maybe an `if isa(nlp, AbstractNLSModel)` could work on that step?

**Comment by dpo** [comment]
Would that be type stable? By the way, it would be great to do the same with TRON.

**Comment by dpo** [comment]
This requires a new release of NLPModels.

**Comment by abelsiqueira** [comment]
Not sure about the type stability. About the TRON LS version, I'll try to take a look soon.

**Comment by dpo** [comment]
We need to think this through because ideally, we should be able to pass in a solver for the subproblem. I'd like to know of good ways to do it. In trunkls, I'm using cgls, but we could pass other solvers that have the right properties. In trunk, tron, tronls, and other solvers, the same situation will happen.

**Comment by dpo** [comment]
Something like https://github.com/mauro3/SimpleTraits.jl perhaps.

**Comment by abelsiqueira** [comment]
Revisiting my answer, I think just `objgrad` won't do. Probably only memoization will prevent full code redundancy.
On the traits, I haven't quite grasped the concept yet. Is it to establish which subproblem solvers work with each type of problem? Like `cgls` for `NLS`, etc.?

**Comment by abelsiqueira** [comment]
It seems memoization is not so difficult to setup, except for the inplace methods. Related: https://github.com/simonster/Memoize.jl/issues/23

**Comment by dpo** [comment]
Where do you think we're performing redundant evaluations here? I was talking about redundant code structure.

Traits could help pass a solver for the trust-region subproblem as argument. For example, we could state that certain solvers are able to solve TR subproblems using something like
```julia
@traitdef CanSolveTRSubproblem{LinearSolver} <- can_solve_tr_subproblem(LinearSolver)
```
If `cg` and `cgls` have a type that derives from `LinearSolver`, we could say
```julia
can_solve_tr_subproblem(CGType) = true
can_solve_tr_subproblem(CGLSType) = true
can_solve_tr_subproblem(GMRESType) = false
```
Then, there is a mechanism to write the `trunk` function to state that the subproblem solver argument must have that trait. Something like
```julia
@traitfn function trunk{LinearSolver; CanSolveTRSubproblem{LinearSolver}}(model::AbstractNLPModel, subsolver::LinearSolver)
  ...
end
```
At least, that's my current understanding of traits.

**Comment by abelsiqueira** [comment]
Sorry, I meant that if we have memoization we can avoid creating additional code by simply using the general version. For instance,
```julia
f = obj(nlp, x)
g = grad(nlp, x)
```
for `nlp` being a `NLS` we have
```julia
function obj(nls, x)
  r = residual(nls, x)
  return 0.5 * dot(r, r)
end
function grad(nls, x)
  r = residual(nls, x) # memoized
  return jtprod(nls, x, r)
end
```

I think I understand traits now, in theory at least.

**Comment by dpo** [comment]
Yes, memoization would definitely help in other places too. Note that currently, `trunkls` evaluates the residual and the Jacobian as an operator, so I don't think it performs redundant evaluations, does it?

**Comment by abelsiqueira** [comment]
No, I don't think it does.

**Comment by dpo** [comment]
The test on 0.7 fail because for some reason we use MathProgBase 0.6.4 instead of 0.7.0. I don't see that a dependency is forcing that version of MPB though, do you?

**Comment by abelsiqueira** [comment]
Nothing forcing 0.6.4, as far as I can tell.

**Comment by dpo** [comment]
Rebased.

@abelsiqueira The tests are going to fail with the message
```
Simple test: Error During Test at /Users/dpo/dev/julia/JSO/Optimize.jl/test/solvers/trunkls.jl:4
  Got exception outside of a @test
  UndefVarError: v not defined
  Stacktrace:
   [1] statsgetfield(::GenericExecutionStats, ::Symbol) at /Users/dpo/dev/julia/JSO/Optimize.jl/src/stats/stats.jl:125
```
Will that be fixed by #77 ?

**Comment by abelsiqueira** [comment]
It appears to be the case. You found a bug in line 125 of stats, where it should say
`Unknown field $name`. So it's supposed to be one of the `neval_` for residuals.

**Comment by dpo** [comment]
Fix it here or in #77?

**Comment by abelsiqueira** [comment]
Here, because we can
- Change the code here and actually see if the error changes to the expected `Unknown field neval_residual`;
- Merge #77;
- Rebased this and see if works then.

**Comment by dpo** [comment]
Seems ok. Now we have
```
Simple test: Error During Test at /home/travis/build/JuliaSmoothOptimizers/Optimize.jl/test/solvers/trunkls.jl:4
  Got exception outside of a @test
  Unknown field neval_residual
  Stacktrace:
   [1] error(::String) at ./error.jl:33
   [2] statsgetfield(::GenericExecutionStats, ::Symbol) at /home/travis/build/JuliaSmoothOptimizers/Optimize.jl/src/stats/stats.jl:125
```

**Comment by abelsiqueira** [comment]
Then we can merge #77.

File: test/solvers/trunkls.jl [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    model = ADNLSModel(x -> [10 * (x[2] - x[1]^2), 1 - x[1]], [-2.0, 1.0], 2)
+    for subsolver in Optimize.trunkls_allowed_subsolvers
+        stats = trunk(model, subsolver=subsolver)
+        @test stats.status == :first_order
+        isapprox(stats.solution, ones(2), rtol=1e-4)
+        @test isapprox(stats.objective, 0, atol=1e-6)
+        @test neval_jac_residual(model) == 0
+        stline = statsline(stats, [:objective, :dual_feas, :elapsed_time, :iter, :neval_residual,
+                                   :neval_jprod_residual, :neval_jtprod_residual, :status])
+        @info @sprintf("%8s  %5d  %5d  %s\n", model.meta.name, nlp.nls_meta.nvar, nlp.nls_meta.nequ, stline)
```
Comment (abelsiqueira) [inline test/solvers/trunkls.jl]: `nlp` -> `model`
It looks like there are some silent errors when we call `@info` and its arguments produces errors: https://travis-ci.org/JuliaSmoothOptimizers/Optimize.jl/jobs/490333777#L875

File: test/solvers/trunkls.jl [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    for subsolver in Optimize.trunkls_allowed_subsolvers
+        stats = with_logger(NullLogger()) do
+            trunk(model, subsolver=subsolver)
+        end
+        @test stats.status == :first_order
+        @test isapprox(stats.objective, 0, atol=1e-6)
+        @test neval_jac_residual(model) == 0
+        stline = statsline(stats, [:objective, :dual_feas, :elapsed_time, :iter, :neval_residual,
+                                   :neval_jprod_residual, :neval_jtprod_residual, :status])
+        @info @sprintf("%8s  %5d  %5d  %s\n", model.meta.name, nlp.nls_meta.nvar, nlp.nls_meta.nequ, stline)
```
Comment (abelsiqueira) [inline test/solvers/trunkls.jl]: same

File: test/solvers/trunkls.jl (hunk 1 above)
Comment (dpo) [reply test/solvers/trunkls.jl]: Good catch!

Comment (dpo) [reply test/solvers/trunkls.jl]: Should be fixed. It's not good that those don't trigger a failure.

**Comment by abelsiqueira** [comment]
Thanks

**Pull request merged**